crates/wassette/src/component_storage.rs

```diff
@@ -0,0 +1,313 @@
+            }
+        }
+        Ok(())
+    }
+
+    /// Create a validation stamp for a component artifact to track stale data on disk.
+    ///
+    /// When `include_hash` is `true` the SHA-256 hash of the file is
+    /// recorded in addition to size and modification time so changes can be
+    /// detected even when timestamps are unreliable.
+    pub async fn create_validation_stamp(
+        &self,
+        path: &Path,
+        include_hash: bool,
+    ) -> Result<ValidationStamp> {
+        let metadata = tokio::fs::metadata(path)
+            .await
+            .with_context(|| format!("Failed to read metadata for {}", path.display()))?;
+
+        let file_size = metadata.len();
+        let mtime = metadata
+            .modified()
+            .map_err(|_| std::io::Error::from(std::io::ErrorKind::Other))
+            .and_then(|t| {
+                t.duration_since(std::time::UNIX_EPOCH)
+                    .map_err(|_| std::io::Error::from(std::io::ErrorKind::Other))
+            })?
+            .as_secs();
+
+        let content_hash = if include_hash {
+            Some(compute_file_hash(path).await?)
+        } else {
+            None
+        };
+
+        Ok(ValidationStamp {
+            file_size,
+            mtime,
+            content_hash,
+        })
+    }
+
+    /// Check if the validation stamp matches the current file on disk.
+    pub async fn validate_stamp(path: &Path, stamp: &ValidationStamp) -> bool {
+        let metadata = match tokio::fs::metadata(path).await {
+            Ok(metadata) => metadata,
+            Err(_) => return false,
+        };
+
+        if metadata.len() != stamp.file_size {
+            return false;
+        }
+
+        if let Some(expected_hash) = &stamp.content_hash {
+            match compute_file_hash(path).await {
+                Ok(actual_hash) => return actual_hash == *expected_hash,
+                Err(_) => return false,
+            }
+        }
+
+        let mtime = match metadata
+            .modified()
+            .map_err(|_| std::io::Error::from(std::io::ErrorKind::Other))
+            .and_then(|t| {
+                t.duration_since(std::time::UNIX_EPOCH)
+                    .map_err(|_| std::io::Error::from(std::io::ErrorKind::Other))
+            })
+            .map(|d| d.as_secs())
+        {
+            Ok(mtime) => mtime,
+            Err(_) => return false,
+        };
+
+        if mtime != stamp.mtime {
+            return false;
+        }
+
+        true
+    }
+}
+
+async fn compute_file_hash(path: &Path) -> Result<String> {
+    let file = tokio::fs::File::open(path)
+        .await
+        .with_context(|| format!("Failed to open {} for hashing", path.display()))?;
+
+    let file = file.into_std().await;
+
+    let path = path.to_path_buf();
+    spawn_blocking(move || -> Result<String> {
+        let mut reader = BufReader::new(file);
+        let mut hasher = Sha256::new();
+        let mut buffer = [0u8; 16 * 1024];
```

Comment:
This should be using `std::io::copy` rather than a manual buffer read (which could be error prone)